### PR TITLE
Add tracking for defensive actions including:

### PR DESCRIPTION
**Pull request by aolowin**
Armour types
  Dodged, Phased, or Deflected attacks
  Blocked and Reflected attacks
Add tracking of auxillary attacks.
Rework count_action() to have a standard way to store auxillary information instead
of doing manual bit shifting.

### CODE REVIEW

**Comment by alexjurkiewicz** [comment]
Nice!

File: crawl-ref/source/actor.cc [hunk 1]
```diff
@@ -368,6 +368,15 @@ int actor::apply_ac(int damage, int max_damage, ac_type ac_rule,
     }
 
     saved = max(saved, min(gdr * max_damage / 100, ac / 2));
+    if (for_real && (damage > 0) && (saved >= damage) && is_player())
+    {
+        const item_def *body_armour = you.slot_item(EQ_BODY_ARMOUR);
```
Comment (wheals) [inline crawl-ref/source/actor.cc]: This is in an actor method, so just `slot_item(EQ_BODY_ARMOUR)` would work

File: crawl-ref/source/player.cc [hunk 2]
```diff
@@ -5676,6 +5676,11 @@ void player::shield_block_succeeded(actor *foe)
 
     shield_blocks++;
     practise(EX_SHIELD_BLOCK);
+    const item_def *shield = you.slot_item(EQ_SHIELD);
```
Comment (wheals) [inline crawl-ref/source/player.cc]: Ditto

**Comment by neilmoore** [comment]
This doesn't handle save compat of old action_counts.  That causes behaviour ranging from wrong labels to crashes when dumping a game that was transferred into this version, because the format of the 32-bit subtype number changed for unarmed, melee, and ranged attacks.

Also, if the subtype is negative, that overwrite the auxtype with -2 when combining the two.

Reverting for now so that servers don't have a hard-to-correct mix of crashy and non-crashy games.

**Comment by neilmoore** [comment]
There doesn't seem to be a way to re-open a merged (as opposed to closed) pull request.  aolowin, could you resubmit this?  I would like to see it go in, it just needs a few fixes and there wasn't enough time to make them before the nightly server rebuilds.

**Pull request merged**